src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1151,11 +1439,47 @@ RebalanceTableShards(RebalanceOptions *options, Oid shardReplicationModeOid)
 }
 
 
+/*
+ * ConflictShardPlacementUpdateOnlyWithIsolationTesting is only useful for
+ * testing and should not be called by any code-path except for
+ * UpdateShardPlacement().
+ *
+ * To be able to test the rebalance monitor functionality correctly, we need to
+ * be able to pause the rebalancer at a specific place in time. We cannot do
+ * this by block the shard move itself someway (e.g. by calling truncate on the
+ * distributed table). The reason for this is that we do the shard move in a
+ * newly opened connection. This causes our isolation tester block detection to
+ * not realise that the rebalance_table_shards call is blocked.
+ *
+ * So instead, before opening a connection we lock an advisory lock that's
+ * based on the shard id. By locking this advisory lock in a different session
+ * we can block the rebalancer in a way that the isolation tester block
+ * detection is able to detect.
+ */
```

Comment:
:sadpanda:

really don't like hacks like this purely for testing. Always wonder if it is worth testing if we need to jump through hoops like this.

Reply:
as per comment below

> This is copied from the enterprise version of citus. This approach is already used there for something similar. I'll make sure this is fixed in the merge.